### DISCUSSION TITLE
Implement a polymorphic method

### DISCUSSION THREAD

**Discussion by gregghz** [scala/scala3#15299] [Metaprogramming]
I've created a macro that takes a type parameter (referencing a trait) and attempts to create an implementation of that trait.

That is, calling: `make[TestTrait]` should produce a class that implements `TestTrait`. However, so far, I am unable to implement polymorphic methods. `TestTrait` has this abstract method: `def poly[A](x: List[A]): A` and I want the implementation created via the macro to just call `x.head`.

My attempt at this is here: https://gist.github.com/gregghz/6b74e50c1600c30322403449ce978a8c

The code produced looks how I'd expect/want with a class defined and an override for poly that looks like this:

```
def poly[A](x: scala.collection.immutable.List[A]): A = x.head
```

However, the compilation fails with an error saying that my class needs to be abstract because `poly` is not defined. The error says that my override does not match because their type parameters differ even though the compiler output shows them side by side and they look identical:

```
[error]   |(Note that
[error]   | parameter List[A] in def poly[A](x: List[A]): A in trait TestTrait in package macros does not match
[error]   | parameter List[A] in def poly[A](x: List[A]): A in class ImplT
[error]   | their type parameters differ)
```

Can anyone point me in the right direction here?

**Comment by bishabosha** [comment]
The issue is that type parameters are unique for each method, so should not be shared. In your example, you are reusing the type for `x` from the original `poly` method, this references the old `A` type parameter from the original `poly`, rather than the new `A` you defined for your new `poly` method.

Currently in `Quotes` there is `substituteTypes` which can substitute references to symbols by types, however I do not think this will work for type parameters of methods - they do not use symbols.

**Reply by gregghz** [reply]
This all makes sense. FWIW, the copying approach was used because in my real use case, I'm listing all methods on the trait and creating stub implementations. That is, I don't know up front the number of and types of the parameters.

How can I reference the A created in the new `poly` method?

```
      val methodType = PolyType(List("A"))(
        _ => List(TypeBounds(TypeRepr.of[Nothing], TypeRepr.of[Any])),
        _ =>
          MethodType(List("x"))(
            mt => List(mt.param(0)),
            _ => abstractDef.returnTpt.tpe
          )
      )
```

Oddly, this ^^ just gets stuck in an infinite loop of some sort. Normally the whole thing compiles in less than 5 seconds. When I use that implementation, it never finishes compiling (at least, I've waiting as long as 5 minutes).
